kubert/src/lease.rs

```diff
@@ -0,0 +1,464 @@
+                        claim: Some(claim.clone()),
+                        meta,
+                    };
+                    return Ok(claim);
+                }
+
+                // The claim is held by another claimant, return it.
+                if claim.is_current() {
+                    return Ok(claim.clone());
+                }
+            }
+
+            // There's no current claim, so try to acquire it.
+            let (claim, meta) = match self.acquire(&state.meta, claimant, params).await {
+                Ok(acquire) => acquire,
+                Err(e) if Self::is_conflict(&e) => {
+                    // Another process updated the claim's resource version, so
+                    // re-sync the state and try again.
+                    *state = Self::get(self.api.clone(), &*self.name).await?;
+                    continue;
+                }
+                Err(e) => return Err(e),
+            };
+            *state = State {
+                claim: Some(claim.clone()),
+                meta,
+            };
+            return Ok(claim);
+        }
+    }
+
+    /// Clear out the state of the lease if the claim is currently held by the
+    /// provided identity.
+    ///
+    /// This is typically used during process shutdown so that another process
+    /// can potentially claim the lease before the prior lease duration expires.
+    pub async fn abdicate(&self, identity: &str) -> Result<bool, Error> {
+        let mut state = self.state.lock().await;
+        if let Some(claim) = state.claim.take() {
+            if claim.is_current_for(identity) {
+                self.patch(&kube_client::api::Patch::Strategic(serde_json::json!({
+                    "apiVersion": "coordination.k8s.io/v1",
+                    "kind": "Lease",
+                    "metadata": {
+                        "resourceVersion": state.meta.version,
+                    },
+                    "spec": {
+                        "acquireTime": Option::<()>::None,
+                        "renewTime": Option::<()>::None,
+                        "holderIdentity": Option::<()>::None,
+                        "leaseDurationSeconds": Option::<()>::None,
+                        // leaseTransitions is preserved by strategic patch
+                    },
+                })))
+                .await?;
+                return Ok(true);
+            }
+        }
+
+        Ok(false)
+    }
+
+    /// Spawn a task that ensures the lease is claimed.
+    ///
+    /// When the lease becomes unclaimed, the task attempts to claim the lease
+    /// as _claimant_ and maintains the lease until the task completes or the
+    /// lease is claimed by another process.
+    ///
+    /// The state of the lease is published via the returned receiver.
+    ///
+    /// When all receivers are dropped, the task completes and the lease is
+    /// abdicated so that another process can claim it.
+    pub async fn spawn_claimant(
+        self,
```

Comment:
I don't think we need to support spawning an arc's lease -- it wouldn't make much sense to have other accessors contending for the lock.

Reply:
hmm...alternatively, we could `Arc` the lease's inner fields so that it's cheap to clone (and the `name` and `field_manager` can be cheaply converted to `Arc<str>`s, so we still only deref one heap pointer to access them), but that would also inflict `Arc`ing on everyone...